File: TTBI_2D/b09_beam_frq.py
import numpy as np
import scipy.linalg as la
import matplotlib.pyplot as plt
import math

# Dummy class for structure initialization if needed
class EmptyObj:
    pass

def b09_beam_frq(Beam, Calc):
    """
    Calculates the beam frequencies given its system matrices
    """
    
    # Initialize Modal if it doesn't exist
    if not hasattr(Beam, 'Modal'):
        Beam.Modal = EmptyObj()

    # Convert sparse matrices to dense arrays for the eigensolver
    # (Matches MATLAB's full() command)
    Kg_dense = Beam.Mesh.Kg.toarray()
    Mg_dense = Beam.Mesh.Mg.toarray()

    # Only natural frequencies calculation
    if Calc.Options.calc_beam_frq == 1 and Calc.Options.calc_beam_modes == 0:
        
        # eigh returns eigenvalues (eigvals_only=True skips eigenvectors)
        # Note: eigh assumes matrices are symmetric/Hermitian. 
        lambda_vals = la.eigh(Kg_dense, Mg_dense, eigvals_only=True)
        
        Beam.Modal.w = np.sqrt(lambda_vals)
        Beam.Modal.f = Beam.Modal.w / (2 * np.pi)

        # Removing values associated to BC
        Beam.Modal.w = Beam.Modal.w[Beam.BC.num_DOF_fixed:]
        Beam.Modal.f = Beam.Modal.f[Beam.BC.num_DOF_fixed:]

    # Natural frequencies and Modes of vibration calculation
    elif Calc.Options.calc_beam_frq == 1 and Calc.Options.calc_beam_modes == 1:
        
        # eigh returns eigenvalues and eigenvectors
        lambda_vals, V = la.eigh(Kg_dense, Mg_dense)
        
        # Sort eigenvalues and eigenvectors (eigh usually sorts them, but doing it to be safe)
        idx = np.argsort(lambda_vals)
        lambda_vals = lambda_vals[idx]
        V = V[:, idx]
        
        # Normalization of eigenvectors
        # V.T @ Mg_dense @ V is the matrix equivalent of V'*Mg*V in MATLAB
        Factor = np.diag(V.T @ Mg_dense @ V)
        Beam.Modal.modes = V / np.sqrt(Factor)
        
        # EigenValues to Natural frequencies
        Beam.Modal.w = np.sqrt(lambda_vals)
        Beam.Modal.f = Beam.Modal.w / (2 * np.pi)
        
        # Removing values associated to BC
        Beam.Modal.w = Beam.Modal.w[Beam.BC.num_DOF_fixed:]
        Beam.Modal.f = Beam.Modal.f[Beam.BC.num_DOF_fixed:]
        # Remove the first 'num_DOF_fixed' columns
        Beam.Modal.modes = Beam.Modal.modes[:, Beam.BC.num_DOF_fixed:]

    # ------------------------------ Plotting ---------------------------------
    
    if not hasattr(Calc.Plot, 'P1_Beam_frq'): Calc.Plot.P1_Beam_frq = 0
    if not hasattr(Calc.Plot, 'P2_Beam_modes'): Calc.Plot.P2_Beam_modes = 0

    # -- Plotting of calculated Natural frequencies --
    if Calc.Plot.P1_Beam_frq == 1:
        plt.figure()
        # range generates integers starting at 1 for the x-axis
        modes_range = list(range(1, len(Beam.Modal.f) + 1))
        plt.plot(modes_range, Beam.Modal.f, '.')
        plt.axis('tight')
        plt.xlabel('Mode number')
        plt.ylabel('Frequency (Hz)')
        plt.title(f"Beam Only (1st frq: {round(Beam.Modal.f[0], 2)} Hz;  Last frq: {round(Beam.Modal.f[-1], 2)} Hz)")
        plt.pause(0.25)

    # -- Plotting Mode shapes --
    
    if Calc.Plot.P2_Beam_modes >= 1:
        num_modes_to_plot = int(Calc.Plot.P2_Beam_modes)
        
        # Calculate rows for subplots
        aux1 = math.ceil(num_modes_to_plot / 2)
        plt.figure()
        
        for k in range(num_modes_to_plot):
            # Extract vertical displacements: 0-indexed, step of 2 -> 0::2
            vert_modes = Beam.Modal.modes[0::2, k]
            aux2 = np.max(np.abs(vert_modes))
            
            plt.subplot(aux1, 2, k + 1) # Subplot index is 1-based in matplotlib
            Xdata = Beam.Mesh.Nodes.acum
            
            plt.plot(Xdata, vert_modes / aux2)
            plt.xlim([0, Beam.Prop.L])
            plt.title(f"Mode {k + 1} ({round(Beam.Modal.f[k], 3)} Hz)")
            
        plt.tight_layout() # Prevents subplot titles from overlapping
        plt.pause(0.25)

    return Beam

# ---- End of function ----

File: scour_MATLAB/B09_BeamFrq.m
function [Beam] = B09_BeamFrq(Beam,Calc)

% Calculates the beam frequencies given its system matrices

% *************************************************************************
% *** Script part of TTB-2D tool for Matlab environment.                ***
% *** Licensed under the GNU General Public License v3.0                ***
% *** Author: Daniel Cantero (daniel.cantero@ntnu.no)                   ***
% *** For help, modifications, and collaboration contact the author.    ***
% *************************************************************************

% -------------------------------------------------------------------------
% ---- Input ----
% Beam = Structure with Beam's variables, including at least:
%   .Kg = Global Stiffness matrix
%   .Mg = Global Mass matrix
%   .BC.num_DOF_fixed = Number of fixed DOF
%   .Prop.L = Beam length
% Calc = Structre with calculation variables. It should include at least:
%   .Options.calc_beam_frq = If value = 1, beam's natural frequencies are calculated
%   .Options.calc_beam_modes = If value = 1, beam's modes of vibration are calculated
% ---- Output ----
% Beam = Addition of fields to structure Beam:
%   .w = Circular frequencies of the beam
%   .f = Natural frequencies of the beam in [Hz]
%   .modes = Modes of vibration of in columns
% -------------------------------------------------------------------------

% Only natural frequencies calculation
if and(Calc.Options.calc_beam_frq == 1,Calc.Options.calc_beam_modes == 0)

    lambda = eig(full(Beam.Mesh.Kg),full(Beam.Mesh.Mg));
    [lambda,~] = local_sorted_eigenvalues( ...
        lambda, Beam.Modal.num_rigid_modes);
    Beam.Modal.w = sqrt(lambda);
    Beam.Modal.f = Beam.Modal.w/(2*pi);

    % Removing values associated to BC
    Beam.Modal.w = Beam.Modal.w(Beam.BC.num_DOF_fixed+1:end);
    Beam.Modal.f = Beam.Modal.f(Beam.BC.num_DOF_fixed+1:end);

% Natural frequencies and Modes of vibration calculation
elseif and(Calc.Options.calc_beam_frq == 1,Calc.Options.calc_beam_modes == 1)
    
    [V,lambda] = eig(full(Beam.Mesh.Kg),full(Beam.Mesh.Mg));
    [lambda,k] = local_sorted_eigenvalues( ...
        diag(lambda), Beam.Modal.num_rigid_modes);
    V = V(:,k); 
    
    % Normaliztion of eigenvectors
    Factor = diag(V'*Beam.Mesh.Mg*V); 
    Beam.Modal.modes = V/(sqrt(diag(Factor)));
    
    % EigenValues to Natural frequencies
    Beam.Modal.w = sqrt(lambda);
    Beam.Modal.f = Beam.Modal.w/(2*pi);
    
    % Removing values associated to BC
    Beam.Modal.w = Beam.Modal.w(Beam.BC.num_DOF_fixed+1:end);
    Beam.Modal.f = Beam.Modal.f(Beam.BC.num_DOF_fixed+1:end);
    Beam.Modal.modes(:,1:Beam.BC.num_DOF_fixed) = [];
    
end % if and(Calc.Options.calc_beam_frq == 1,Calc.Options.calc_beam_frq == 0)

% ------------------------------ Plotting ---------------------------------
    
% -- Plotting of calculated Natural frequencies --
if Calc.Plot.P1_Beam_frq == 1
    figure; plot((1:length(Beam.Modal.f)),Beam.Modal.f,'.'); axis tight;
    %figure; semilogy((1:length(Beam.Modal.f)),Beam.Modal.f,'.'); axis tight;
    xlabel('Mode number'); ylabel('Frequency (Hz)');
    title(['Beam Only (1st frq: ',num2str(round(Beam.Modal.f(1),2)),' Hz;',...
        blanks(1),'Last frq: ',num2str(round(Beam.Modal.f(end),2)),' Hz)']);
    pause(0.25);
end % if Calc.Plot.P1_Beam_frq == 1

% -- Plotting Mode shapes --
if Calc.Plot.P2_Beam_modes >= 1
    if Calc.Plot.P2_Beam_modes > 0
        aux1 = ceil(Calc.Plot.P2_Beam_modes/2);
        figure; 
        for k = 1:Calc.Plot.P2_Beam_modes
            aux2 = max(abs(Beam.Modal.modes(1:2:end,k)));
            subplot(aux1,2,k)
            Xdata = Beam.Mesh.Nodes.acum;
            plot(Xdata,Beam.Modal.modes(1:2:end,k)/aux2);
            xlim([0,Beam.Prop.L]);
            title(['Mode ',num2str(k),' (',num2str(round(Beam.Modal.f(k),3)),' Hz)']);
        end % for k = 1:Calc.Plot.P2_Beam_modes
        pause(0.25);
    end % if Calc.Plot.P2_Beam_modes > 0
end % if Calc.Plot.P2_Beam_modes == 1

% ---- End of function ----
end

function [lambda, order] = local_sorted_eigenvalues( ...
        raw_lambda, expected_rigid_modes)
% Generalized EIG does not promise eigenvalue order.  Rayleigh damping and
% fixed-DOF removal both require ascending modal order, so make it explicit
% and reject anything beyond round-off departure from a real PSD spectrum.
if ~isscalar(expected_rigid_modes) || ...
        expected_rigid_modes ~= floor(expected_rigid_modes) || ...
        expected_rigid_modes < 0 || expected_rigid_modes > 2
    error('B09:InvalidRigidModeCount', ...
        'Expected beam rigid-mode count must be an integer in [0,2].');
end
if any(~isfinite(raw_lambda))
    error('B09:NonfiniteEigenvalue', ...
        'Generalized beam eigenproblem returned a non-finite eigenvalue.');
end
scale = max([1; abs(raw_lambda(:))]);
% A global relative tolerance such as 1e-10*max(lambda) is far too loose for
% FE spectra spanning many decades.  Bound round-off by 128 ULPs at the
% largest eigenvalue, then require exactly the independently derived number
% of near-zero rigid modes.
tol = 128 * eps(max(scale));
if any(abs(imag(raw_lambda(:))) > tol)
    error('B09:ComplexEigenvalue', ...
        'Generalized beam eigenproblem returned a materially complex value.');
end
lambda = real(raw_lambda(:));
if any(lambda < -tol)
    error('B09:NegativeEigenvalue', ...
        'Generalized beam eigenproblem returned a materially negative value.');
end
[lambda, order] = sort(lambda, 'ascend');
near_zero = abs(lambda) <= tol;
if sum(near_zero) ~= expected_rigid_modes
    error('B09:RigidModeCountMismatch', ...
        ['Eigenproblem has %d near-zero modes within %.6g, but boundary ' ...
         'conditions predict %d.'], ...
        sum(near_zero), tol, expected_rigid_modes);
end
lambda(near_zero) = 0;
end
